File: Java/Leetcode/Stack/AsteroidCollision.java
package com.brianprojects.Leetcode.Stack;

import java.util.Stack;

public class AsteroidCollision {
    public int[] asteroidCollision(int[] asteroids){
        if(asteroids == null){
            return null;
        }
        Stack<Integer> stack = new Stack<>();

        for(var num: asteroids){
            if(num<0){
                while(!stack.isEmpty() && stack.peek()>0&& Math.abs(num) > stack.peek()){
                    stack.pop();
                }
                if(stack.isEmpty() || stack.peek() < 0){
                    stack.push(num);
                }
                if(!stack.isEmpty() && Math.abs(num) == stack.peek()){
                    stack.pop();
                }
            }
            else{
                stack.push(num);
            }
        }

        int[] res = new int[stack.size()];

        for(var i=res.length-1; i>=0; i--){
            res[i] = stack.pop();
        }
        return res;
    }
}

File: Python/Leetcode_Python/Stack/AsteroidCollision.py
# Time Complexity: O(n) || Space Complexity: O(n)
from typing import List


def asteroidCollision(asteroids: List[int]) -> List[int]:
    if not asteroids:
        return []

    stack = []

    for asteroid in asteroids:
        if not stack:
            stack.append(asteroid)
            continue

        if stack[-1] < 0 or (stack[-1] > 0 and asteroid > 0):
            stack.append(asteroid)
            continue

        if stack and abs(stack[-1]) == abs(asteroid):
            stack.pop()
            continue

        while (
            stack and asteroid < 0 and stack[-1] > 0 and abs(asteroid) > abs(stack[-1])
        ):
            stack.pop()

        if stack and asteroid < 0 and stack[-1] > 0 and abs(stack[-1]) == abs(asteroid):
            stack.pop()
        elif not stack or (
            stack and stack[-1] // abs(stack[-1]) == asteroid // abs(asteroid)
        ):
            stack.append(asteroid)

    return stack
